### DISCUSSION TITLE
DAQs and syncing

### DISCUSSION THREAD

**Discussion by emikrish** [sn-lab/MouseGoggles#28] [General]
Hello!

I was wondering if you have any thoughts on how to sync external data acquisition devices (like fiber photometry, even video recordings) to the Godot behavioral logs. I have tried processing TTL pulses from a DAQ through an Arduino keyboard emulator, so that Godot can log 'keypresses' as events. But I am finding that this method isn't very reliable. Has anyone encountered a similar issue or explored other solutions for syncing external devices with Godot?  I’d appreciate any ideas!!

Thanks! :)

**Comment by misaacson01** [comment]
For synchronization, you could go in either direction: either log external events into Godot, or log Godot events to some external DAQ. To keep things simple, I would recommend basically the same thing you described: log external using a keyboard or gamepad emulator mapping those events as key or button presses. That's seemed to work well for us, but I'd be curious to hear what problems you're running into.

One thing that comes to mind is that since Godot can only log things as fast as the game updates (e.g. ~60 fps), having Godot handle the logging only makes sense if the events you're logging are much more sparse than game updates. If the events are close to or faster than this (e.g. video frames), than I think it makes sense to log this in the opposite direction, maybe even having a separate DAQ dedicated to logging all events at a fast sample rate. As an example of this external logging, we've used basically a similar concept as used for Godot logging -- have Godot send an external gamepad command that is read by a gamepad emulator. It's funny, but the only built-in function I've found for sending data out from Godot (other than the video signal, obviously) is "vibration" commands for gamepads with rumble packs. Sending a vibration command essentially gives a way to send out some bytes over USB which can be read by a gamepad emulator. This is a little more complicated than emulating a keyboard since microcontrollers have to have the right chip to be able to emulate a gamepad, but once it's set up it works very well. We used this method to send commands for liquid reward delivery during the linear track experiment (for example, see this [linear track godot script](https://github.com/sn-lab/MouseGoggles/blob/main/Godot/MouseVR%20Godot%20Project%20V1.8/linearTrackScene.gd), line 172). If you're interested in going this direction, I can add some more details to the repository for how to set up this gamepad emulator for reading Godot output data.

**Reply by sbugeon** [reply]
Hello,

Question related to this discussion about syncing 

I am using the EyeTrack 1.1, and I would like to be able to send a TTL pulse from the Raspberry Pi 4 to a DAQ card, to log as precisely as possible when stimuli appear on the screen.

Is there a way to send the TTL pulse directly from GPIO pins?  (there are some remaining pins after pluging the screens, and LEDs but not sure if they can be used for this?)

In Godot, I think you wrote a function to do this (pulse_output), is that correct? 

Thanks!

**Reply by misaacson01** [reply]
Hello @sbugeon 

Yes, the "pulse_output" function in Godot does send a TTL pulse like you're asking about. However, this pulse is output from the Raspberry Pi 5 (the one connected to the eye-tracking cameras), not the Pi 4. I did this because there was already a convenient communication channel (TCP) between the Pi 4 and Pi 5 to transfer eye-tracking images, and the Pi 5 has all of its pins exposed. Is this what you're looking for? 

It might be possible to send a TTL pulse from Godot out of the Pi 4 directly if that's what you need, but I haven't tested this yet (earlier versions couldn't do it due to a locked-down Godot version, but the new version might be able to).

**Reply by sbugeon** [reply]
Ah I see, that's a good idea, I could use the TCP communication to send a signal to the Pi 5, which will be sending the TTL pulse to the DAQ card. 

Many thanks!
